### DISCUSSION TITLE
No OnPropertChanged event raised for a defined AvaloniaProperty when Datacontext is set

### DISCUSSION THREAD

**Discussion by msgan** [AvaloniaUI/Avalonia#13091] [Q&A] [answered]
Greetings, a newbie to XAML things. I created a view that will be embedded in another with a thought that I could define an AvaloniaProperty for this subview while also having its own Datacontext at the same time. But what is happening is; no OnPropertyChanged call gets ever made for the defined property and never gets data from the containing view when the subview has its own Datacontext.

Is this the expected behavior? Shouldn't a view with own Datacontext also expect to receive variables via defined properties?

    <RatingControl Value="2"/>
    No use setting Value here, it always remains the default 0

    class RatingControlDataContext
    {
        
    }

    public class RatingControl : UserControl 
    {

        public RatingControl()
        {
            DataContext = new RatingControlDataContext(); //When I set this,
            ValueProperty.Changed.Subscribe(Console.WriteLine); //Value remains 0 and this subscription get nothing
        }

        // Or
        protected override void OnPropertyChanged(AvaloniaPropertyChangedEventArgs change)
        {
            base.OnPropertyChanged(change);
            if (change.Property == ValueProperty)
            {
                // Never reached. No such ValueProperty change event ever occurs
            }
        }

        public static readonly DirectProperty<RatingControl, int> ValueProperty =
            AvaloniaProperty.RegisterDirect<RatingControl, int>(
                nameof(Value),
                o => o.Value,         
                (o, v) => o.Value = v,
                defaultBindingMode: BindingMode.TwoWay);

        private int _value;

        public int Value
        {
            get { return _value; }
            set { SetAndRaise(ValueProperty, ref _value, value); }
        }

        //....
        //....
        //....
    }

**Comment by timunie** [comment]
are you looking for a control like this? https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/CustomControls/RatingControlSample

**Reply by msgan** [reply]
I'm not looking for a control specifically, you haven't noticed but that sample is where I took my snippet from, though I only used it to give example. My question is on the possible/intended usage scenarios for defining AvaloniaProperty on user's controls. I think I observe that: such controls with defined property that also have their own DataContext do not receive values bounded to them from their parent's DataContext. WAIT, a more complete example will better show:

[Avapp.zip](https://github.com/AvaloniaUI/Avalonia/files/12807037/Avapp.zip)

I have MainWindow with MainWindowDataContext and I have two other controls, Control1 and Control2, identical to eachother except that Control2's DataContext is set to Control2DataContext. These two each have defined property ValueProperty, whose change is subscribed to from their respective ctors that tracks it. 

    public Control2()
    {
        InitializeComponent();
        DataContext = new Control2DataContext();// not in Control1

        ValueProperty.Changed.Subscribe(
            args => Console.WriteLine("Control2 receives: " + args.NewValue));
    }

MainWindow.axaml looks like:

      <Window  ...
            xmlns:app="using:Avapp"
            x:DataType="app:MainWindowDataContext">
        <StackPanel>
          <app:Control1 Value="{Binding Val}"/>
          <app:Control2 Value="{Binding Val}"/>
  
          
        </StackPanel>
      </Window>

The Console after running the app looks like:

Control1 receives: 1
Control1 receives: 2

My expectation was both Controls would receive the AvaloniaProperty values from MainWindow's DataContext; thinking could use it to create sub-views. So question is whether there is something I'm missing or the result I'm seeing is by design and I should revise how I should go about it.

**Reply by timunie** [reply, marked as answer]
Nah, you set another DataContext for Control2, which has no property called `Val`. so you can't bind to `Val` here. If you cancel this line, it works: 

```diff
    public Control2()
    {
        InitializeComponent();
-       DataContext = new Control2DataContext();

        ValueProperty.Changed.Subscribe(
            args => Console.WriteLine("Control2 receives: " + args.NewValue));
    }
```

then I get this output
```
Control1 receives: 1
Control2 receives: 1
Control1 receives: 2
Control2 receives: 2
```

**Reply by msgan** [reply]
But that's the thing, Why shouldn't I be able to set context for Control2?    Control2 is not trying to reference and bind it for its child controls in its own XAML. It's just expecting to receive it from it's parent control MainWindow.    It might be because I come from a different UI realm and don't know the internals here, that I don't understand, but as a working principle I haven't read anything that relates a control's DataContext to its own AvaloniaProperty.

**Reply by timunie** [reply]
https://docs.avaloniaui.net/docs/next/basics/data/data-binding/data-context

**Reply by msgan** [reply]
Now that I think about it, it's the child view's DataContext that I should want to be updated. So for my scenario; where a complex child view needs its own DataContext and also gets update from from the parent's DataContext; the solution I took is to have the reference of the child's context in the parent's.
**It works, but is it a proper way of doing it?**

MainWindowDataContext.cs

    public class MainWindowDataContext : ReactiveObject
    {
        // [Reactive] public int Val { get; set; } = 1;
        public Control2DataContext Control2DataContext { get; set; }

        public MainWindowDataContext()
        {
            Control2DataContext = new() { Val = 1 };
            UpdateVal();
        }

        async void UpdateVal()
        {
            await Task.Delay(2000);
            Control2DataContext.Val = 2;
        }
    }

Control2DataContext.cs

    public class Control2DataContext : ReactiveObject
    {
       [Reactive] public int Val { get; set; }
    }

MainWindow.axaml

    <app:Control2 DataContext="{Binding Control2DataContext}"/>

**[unnecessary detail]**
What I know and took conceptual correspondence from is ReactJs. If you know it, I imagined AvaloniaProperty as React component Props and DataContext as component State. Of course this parallelism will reach its limits and one of those is what I'm seeing now and had asked about. In React, every component has its own State and also can receive Props from their parent that could dynamically be updated depending on the parent's State. Also being led to think that I've set the property in MainWindow.xaml with no issues, yet getting no effect didn't help my misdirection either.

**Reply by timunie** [reply]
in an async void, you may need to use `Dispater.UiThread.Post` as events only should happen un the UI-thread. Other than that, it should normally work. 

> **Note** 
> I prefer to use the MVVM Community Toolkit as it has source generators which are for me better than the Fody stuff nowadays.
